imageio/plugins/pyav.py

```diff
@@ -0,0 +1,839 @@
+    if subsampled_width or subsampled_height or unaligned_components:
+        raise IOError(
+            f"{frame.format.name} can't be expressed as a strided array."
+            "Use `format=` to select a format to convert into."
+        )
+
+    shape = [frame.height, frame.width]
+
+    n_planes = max([component.plane for component in frame.format.components]) + 1
+    if n_planes > 1:
+        shape = [n_planes] + shape
+
+    channels_per_plane = [0] * n_planes
+    for component in frame.format.components:
+        channels_per_plane[component.plane] += 1
+    n_channels = max(channels_per_plane)
+
+    if n_channels > 1:
+        shape = shape + [n_channels]
+
+    return tuple(shape)
+
+
+class PyAVPlugin(PluginV3):
+    """Support for pyAV as backend.
+
+    Parameters
+    ----------
+    request : iio.Request
+        A request object that represents the users intent. It provides a
+        standard interface to access various the various ImageResources and
+        serves them to the plugin as a file object (or file). Check the docs for
+        details.
+
+    """
+
+    def __init__(
+        self,
+        request: Request,
+        *,
+        container: str = None,
+    ) -> None:
+        """Initialize a new Plugin Instance.
+
+        See Plugin's docstring for detailed documentation.
+
+        Notes
+        -----
+        The implementation here stores the request as a local variable that is
+        exposed using a @property below. If you inherit from PluginV3, remember
+        to call ``super().__init__(request)``.
+
+        """
+
+        super().__init__(request)
+
+        self._container = None
+        self._video_stream = None
+
+        if request.mode.io_mode == IOMode.read:
+            try:
+                self._container = av.open(request.get_file())
+                self._video_stream = self._container.streams.video[0]
+                self._decoder = self._container.decode(video=0)
+            except av.AVError:
+                if isinstance(request.raw_uri, bytes):
+                    msg = "PyAV does not support these `<bytes>`"
+                else:
+                    msg = f"PyAV does not support `{request.raw_uri}`"
+                raise InitializationError(msg)
+        else:
+            file_handle = self.request.get_file()
+            filename = getattr(file_handle, "name", None)
+            extension = self.request.extension or self.request.format_hint
+            if extension is None:
+                raise InitializationError("Can't determine output container to use.")
+
+            # hacky, but beats running our own format selection logic
+            # (since av_guess_format is not exposed)
+            try:
+                setattr(file_handle, "name", filename or "tmp" + extension)
+            except AttributeError:
+                pass  # read-only, nothing we can do
+
+            try:
+                self._container = av.open(file_handle, mode="w", format=container)
```

Comment:
Can we avoid opening twice?

Reply:
Oh? Where is the other open?

Reply:
A few lines up where we establish whether we can read the file.

Reply:
Ah. The open here is for writing and the open call above for reading. Do you mean to try and combine the reading and writing path where possible?